ext/MatrixAlgebraKitMooncakeExt/MatrixAlgebraKitMooncakeExt.jl

```diff
@@ -0,0 +1,225 @@
+module MatrixAlgebraKitMooncakeExt
+
+using Mooncake
+using Mooncake: DefaultCtx, CoDual, Dual, NoRData, rrule!!, frule!!, arrayify, @is_primitive
+using MatrixAlgebraKit
+using MatrixAlgebraKit: inv_safe, diagview, copy_input
+using MatrixAlgebraKit: qr_pullback!, lq_pullback!
+using MatrixAlgebraKit: qr_null_pullback!, lq_null_pullback!
+using MatrixAlgebraKit: eig_pullback!, eigh_pullback!
+using MatrixAlgebraKit: left_polar_pullback!, right_polar_pullback!
+using LinearAlgebra
+
+
+@is_primitive Mooncake.DefaultCtx Mooncake.ReverseMode Tuple{typeof(copy_input), Any, Any}
+function Mooncake.rrule!!(::CoDual{typeof(copy_input)}, f_df::CoDual, A_dA::CoDual)
+    Ac = copy_input(Mooncake.primal(f_df), Mooncake.primal(A_dA))
+    dAc = Mooncake.zero_tangent(Ac)
+    function copy_input_pb(::Mooncake.NoRData)
+        Mooncake.increment!!(Mooncake.tangent(A_dA), dAc)
+        return Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData()
+    end
+    return CoDual(Ac, dAc), copy_input_pb
+end
+
+# two-argument factorizations like LQ, QR, EIG
+for (f, pb, adj) in (
+        (qr_full!, qr_pullback!, :dqr_adjoint),
+        (qr_compact!, qr_pullback!, :dqr_adjoint),
+        (lq_full!, lq_pullback!, :dlq_adjoint),
+        (lq_compact!, lq_pullback!, :dlq_adjoint),
+        (eig_full!, eig_pullback!, :deig_adjoint),
+        (eigh_full!, eigh_pullback!, :deigh_adjoint),
+        (left_polar!, left_polar_pullback!, :dleft_polar_adjoint),
+        (right_polar!, right_polar_pullback!, :dright_polar_adjoint),
+    )
+
+    @eval begin
+        @is_primitive Mooncake.DefaultCtx Mooncake.ReverseMode Tuple{typeof($f), Any, Tuple{<:Any, <:Any}, MatrixAlgebraKit.AbstractAlgorithm}
+        function Mooncake.rrule!!(::CoDual{typeof($f)}, A_dA::CoDual, args_dargs::CoDual, alg_dalg::CoDual{<:MatrixAlgebraKit.AbstractAlgorithm}; kwargs...)
+            A, dA = arrayify(A_dA)
+            args = Mooncake.primal(args_dargs)
+            dargs = Mooncake.tangent(args_dargs)
+            arg1, darg1 = arrayify(args[1], dargs[1])
+            arg2, darg2 = arrayify(args[2], dargs[2])
+            Ac = copy(A)
+            arg1c = copy(arg1)
+            arg2c = copy(arg2)
+            output = $f(A, args, Mooncake.primal(alg_dalg); kwargs...)
+            function $adj(::Mooncake.NoRData)
+                dAtmp_ = zero(dA)
+                dAtmp_ .= $pb(dAtmp_, A, (arg1, arg2), (darg1, darg2); kwargs...)
+                dAtmp = if eltype(dA) <: Real
+                    dAtmp_
+                else
+                    map(A_ -> Mooncake.build_tangent(typeof(A_), real(A_), imag(A_)), dAtmp_)
+                end
+                Mooncake.increment!!(Mooncake.tangent(A_dA), dAtmp)
+                arg1 .= arg1c
+                arg2 .= arg2c
+                A .= Ac
+                darg1 .= zero(darg1)
+                darg2 .= zero(darg2)
```

Comment:
```suggestion
                MatrixAlgebraKit.zero!(darg1)
                MatrixAlgebraKit.zero!(darg2)
```
Is the other implementation not making copies? I might be wrong about that though.